src/service/ampdoc-impl.js

```diff
@@ -0,0 +1,207 @@
+   *
+   * @param {!Node} node
+   * @return {!AmpDoc}
+   */
+  getAmpDoc(node) {
+    // Single document: return it immediately.
+    if (this.singleDoc_) {
+      return this.singleDoc_;
+    }
+
+    // A custom element may already have the reference to the ampdoc.
+    if (typeof node.getAmpDoc == 'function') {
+      const ampDoc = node.getAmpDoc();
+      if (ampDoc) {
+        return ampDoc;
+      }
+    }
+
+    // Otherwise discover and possibly create the ampdoc.
+    // TODO(dvoytenko): Replace with `getRootNode()` API when it's available.
+    const shadowRoot = closestNode(node, node => isShadowRoot(node));
+    if (!shadowRoot) {
+      throw dev.createError('No root found for', node);
+    }
+    let ampDoc = shadowRoot[AMPDOC_PROP];
+    if (!ampDoc) {
+      ampDoc = new AmpDocShadow(shadowRoot);
+      shadowRoot[AMPDOC_PROP] = ampDoc;
+    }
+    return ampDoc;
+  }
+}
+
+
+/**
+ * This class represents a single ampdoc. `AmpDocService` can contain only one
+ * global ampdoc or multiple, depending on the runtime mode: single-doc or
+ * shadow-doc.
+ * @abstract
+ * @package
+ */
+export class AmpDoc {
+
+  /**
+   * @return {!Window}
+   */
+  getWin() {
+    return dev.assert(null, 'not implemented');
+  }
+
+  /**
+   * Returns the root node for this ampdoc. It will either be a `Document` for
+   * the single-doc runtime mode, or a `ShadowRoot` for shadow-doc mode. This
+   * node can be used, among other things, to add ampdoc-wide event listeners.
+   *
+   * @return {!Document|!ShadowRoot}
+   */
+  getRootNode() {
+    return dev.assert(null, 'not implemented');
+  }
+
+  /**
+   * Locates an element with the specified ID within the ampdoc. In the
+   * shadow-doc mode, when multiple documents could be present, this method
+   * localizes search only to the DOM subtree specific to this ampdoc.
+   *
+   * @param {string} id
+   * @return {?Element}
+   */
+  getElementById(id) {
+    return this.getRootNode().getElementById(id);
+  }
+}
+
+
+/**
+ * The version of `AmpDoc` in the single-doc mode that corresponds to the
+ * global `window.document`.
+ * @package @visibleForTesting
+ */
+export class AmpDocSingle extends AmpDoc {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super();
+    /** @private @const {!Window} */
+    this.win = win;
+  }
+
+  /** @override */
+  getWin() {
+    return this.win;
+  }
+
+  /** @override */
+  getRootNode() {
+    return this.win.document;
+  }
+}
+
+
+/**
+ * The version of `AmpDoc` in the shadow-doc mode that is allocated for each
+ * ampdoc hosted within a shadow root.
+ * @package @visibleForTesting
+ */
+export class AmpDocShadow extends AmpDoc {
+  /**
+   * @param {!ShadowRoot} shadowRoot
+   */
+  constructor(shadowRoot) {
+    super();
+    /** @private @const {!ShadowRoot} */
+    this.shadowRoot_ = shadowRoot;
+  }
+
+  /** @override */
+  getWin() {
+    return this.shadowRoot_.ownerDocument.defaultView;
```

Comment:
Our programs are pretty sensitive to the performance of `getWin()` everywhere. Might be worth caching in the constructor.